### PR TITLE
feat: add the support of fsdp2 and remove deepspeed (new version of PR 1115)

### PR DESCRIPTION
**Pull request by LYMDLUT**

### CODE REVIEW

**Comment by MagellaX** [comment]
What do you want me to do in this PR??
Let me know what changes I can make or where to focus!!

File: openrlhf/cli/train_dpo.py [hunk 1]
```diff
@@ -18,19 +19,16 @@ def train(args):
     strategy.setup_distributed()
 
     # configure model
-    # load huggingface model
     model = Actor(
         args.pretrain,
         attn_implementation=args.attn_implementation,
-        param_dtype=args.param_dtype,  # default: bf16
-        load_in_4bit=args.load_in_4bit,
+        torch_dtype=convert_to_torch_dtype("fp32"),
```
Comment (hijkzzz) [inline openrlhf/cli/train_dpo.py]: "fp32": as default value and use torch.float32

File: openrlhf/cli/train_dpo.py [hunk 2]
```diff
@@ -56,6 +50,16 @@ def train(args):
             gradient_checkpointing_kwargs={"use_reentrant": args.gradient_checkpointing_use_reentrant}
         )
 
+    # Wrap/shard model(s) before building optimizer/scheduler (params become DTensor/sharded).
+    model = strategy.prepare(model)
+    # Reference model is inference-only; cpu_offload controls FSDP2 CPUOffloadPolicy.
+    ref_model = strategy.prepare(ref_model, cpu_offload=args.ref_offload)
+    strategy.load_pretrained(model, args.pretrain)
```
Comment (hijkzzz) [inline openrlhf/cli/train_dpo.py]: strategy.load_pretrained_weights?

File: openrlhf/cli/train_dpo.py [hunk 3]
```diff
@@ -125,17 +129,14 @@ def train(args):
...
     )
 
-    # strategy prepare
-    ((model, optim, scheduler), ref_model) = strategy.prepare((model, optim, scheduler), ref_model)
-
     # load checkpoint
     consumed_samples = 0
-    if args.load_checkpoint and os.path.exists(args.ckpt_path):
-        _, states = strategy.load_ckpt(model.model, args.ckpt_path)
+    if args.load_checkpoint and os.path.exists(args.dcp_ckpt_path):
```
Comment (hijkzzz) [inline openrlhf/cli/train_dpo.py]: --checkpoint_path
--save_hf_checkpoint
--disable_fsdp_checkpoint

File: openrlhf/cli/train_dpo.py [hunk 4]
```diff
@@ -203,19 +212,35 @@ def train(args):
...
         default="flash_attention_2",
         help="Attention implementation (e.g., eager, flash_attention_2, flash_attention_3, kernels-community/vllm-flash-attn3)",
     )
     parser.add_argument("--use_liger_kernel", action="store_true", default=False, help="Enable Liger Kernel")
-    parser.add_argument("--grad_accum_dtype", type=str, default=None, help="Adam grad accum data type")
-    parser.add_argument("--overlap_comm", action="store_true", default=False)
     parser.add_argument("--gradient_checkpointing_use_reentrant", action="store_true", default=False)
-    parser.add_argument("--ds_tensor_parallel_size", type=int, default=1, help="DeepSpeed Tensor parallel size")
+    parser.add_argument("--fsdp2_tp_size", type=int, default=1, help="FSDP2 tensor parallel size")
+    parser.add_argument(
```
Comment (hijkzzz) [inline openrlhf/cli/train_dpo.py:249]: fsdp2_sp_size fsdp2 .xxxxx

File: openrlhf/cli/train_dpo.py [hunk 5]
```diff
@@ -203,19 +212,35 @@ def train(args):
...
+        "--sequence_parallel",
+        action="store_true",
+        default=False,
+        help="Enable sequence parallelism for FSDP2+TP (requires --fsdp2_tp_size > 1).",
+    )
+    parser.add_argument(
```
Comment (hijkzzz) [inline openrlhf/cli/train_dpo.py:255]: default

File: openrlhf/cli/train_dpo.py [hunk 6]
```diff
@@ -203,19 +212,35 @@ def train(args):
...
+        "--tp_shard_logits",
+        action="store_true",
+        default=False,
+        help="Shard lm_head logits on vocab dim (FSDP2+TP).",
+    )
+
+    # HF safetensors sharded export (DCP writer)
+    parser.add_argument(
+        "--hf_max_shard_size_gb",
```
Comment (hijkzzz) [inline openrlhf/cli/train_dpo.py:264]: max_checkpoint memory

File: openrlhf/cli/train_dpo.py [hunk 7]
```diff
@@ -203,19 +212,35 @@ def train(args):
...
-    parser.add_argument("--ds_tensor_parallel_size", type=int, default=1, help="DeepSpeed Tensor parallel size")
+    parser.add_argument("--fsdp2_tp_size", type=int, default=1, help="FSDP2 tensor parallel size")
+    parser.add_argument(
+        "--sequence_parallel",
+        action="store_true",
+        default=False,
+        help="Enable sequence parallelism for FSDP2+TP (requires --fsdp2_tp_size > 1).",
+    )
+    parser.add_argument(
+        "--tp_shard_logits",
```
Comment (hijkzzz) [inline openrlhf/cli/train_dpo.py]: shard_vocabury xxxx

File: openrlhf/models/loss.py [hunk 8]
```diff
@@ -19,38 +27,27 @@ def __init__(self, ring_attn_group=None):
...
-        else:
-            shift_logits = logits[..., :-1, :].contiguous()
-            shift_labels = labels[..., 1:].contiguous()
+        shift_logits = logits[..., :-1, :].contiguous()
+        shift_labels = labels[..., 1:].contiguous()
 
-            loss = self.loss(shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1))
-
-        return loss
+        if torch.all(shift_labels == self.IGNORE_INDEX):
```
Comment (hijkzzz) [inline openrlhf/models/loss.py:69]: review ?

Comment (hijkzzz) [reply openrlhf/models/loss.py:69]: delete?

File: openrlhf/trainer/ray/ppo_actor.py [hunk 9]
```diff
@@ -457,34 +471,40 @@ def init_model_from_pretrained(self, strategy: DeepspeedStrategy, pretrain, max_
...
+        self.ema_model = ema_model
 
         # load checkpoint
         self.checkpoint_states = {}
-        ckpt_path = os.path.join(args.ckpt_path, "_actor")
+        ckpt_path = os.path.join(args.dcp_ckpt_path, "_actor")
         if args.load_checkpoint and os.path.exists(ckpt_path):
             strategy.print(f"Loading the checkpoint: {ckpt_path}")
-            _, states = strategy.load_ckpt(self.actor.model, ckpt_path)
+            ckpt_id, states = strategy.load_dcp_model(
```
Comment (hijkzzz) [inline openrlhf/trainer/ray/ppo_actor.py]: load_dcp_chekpoint

File: openrlhf/trainer/sft_trainer.py [hunk 10]
```diff
@@ -205,13 +205,21 @@ def save_logs_and_checkpoints(self, args, global_step, step_bar, logs_dict={}, c
...
+                    args.max_ckpt_num,
+                    args.max_ckpt_mem,
+                    client_states,
+                    optimizer=self.optimizer,
+                    scheduler=self.scheduler,
                 )
             if self.save_hf_ckpt:
-                save_path = os.path.join(args.ckpt_path, f"{tag}_hf")
-                self.strategy.save_model(self.model, self.tokenizer, save_path)
+                self.strategy.save_hf_ckpt_with_rotation(
```
Comment (hijkzzz) [inline openrlhf/trainer/sft_trainer.py]: delete _with_rotation

File: openrlhf/utils/fsdp2/tp/tp_parallel.py [hunk 11]
```diff
@@ -0,0 +1,402 @@
...
+    plan["model.layers.*.self_attn.k_norm"] = _head_shard_norm_style()
+    return plan
+
+
+_MODEL_PLANS = {
+    "LlamaForCausalLM": _llama_plan,
+    "LlamaForSequenceClassification": _llama_plan,
+    "MistralForCausalLM": _llama_plan,
+    "Qwen2ForCausalLM": _qwen_plan,
+    "Qwen3ForCausalLM": _qwen_plan,
```
Comment (hijkzzz) [inline openrlhf/utils/fsdp2/tp/tp_parallel.py]: only supports qwen3 3.5

File: openrlhf/utils/fsdp2/tp/tp_parallel.py [hunk 12]
```diff
@@ -0,0 +1,402 @@
+"""
```
Comment (hijkzzz) [inline openrlhf/utils/fsdp2/tp/tp_parallel.py:1]: too complex
move model tp-related utils to this file

File: openrlhf/utils/fsdp2/checkpoint.py [hunk 13]
```diff
@@ -0,0 +1,805 @@
+"""
```
Comment (hijkzzz) [inline openrlhf/utils/fsdp2/checkpoint.py:1]: check again

Comment (hijkzzz) [reply openrlhf/utils/fsdp2/checkpoint.py:1]: remove sft lora  / all lora related things

File: openrlhf/cli/train_ppo_ray.py [hunk 14]
```diff
@@ -1,4 +1,5 @@
 import argparse
+import os
```
Comment (hijkzzz) [inline openrlhf/cli/train_ppo_ray.py]: combine prompt_len max_new_token len

File: openrlhf/models/utils.py [hunk 15]
```diff
@@ -1,7 +1,280 @@
 from typing import Optional, Tuple, Union
 
 import torch
+import torch.distributed as dist
+import torch.distributed.nn.functional as dist_nn
 import torch.nn.functional as F
+from torch.distributed.tensor import DTensor
+
+
+# ============================================================
```
Comment (hijkzzz) [inline openrlhf/models/utils.py]: comments

**Comment by MagellaX** [comment]
hey can I get the permission to make changes here @LYMDLUT 
i don't have the write access

**Comment by LYMDLUT** [comment]
> hey can I get the permission to make changes here @LYMDLUT i don't have the write access

https://github.com/LYMDLUT/OpenRLHF

**Comment by MagellaX** [comment]
Hey @LYMDLUT, btw, can you tell me what GPUs and how many you are using, and the env??(for testing)

**Comment by MagellaX** [comment]
> Hey @LYMDLUT, btw, can you tell me what GPUs and how many you are using, and the env??(for testing)

hey??

**Comment by jiosephlee** [comment]
@LYMDLUT is there a reason we need to remove support for deepspeed completely?
